File: Python/lirst-fast/lirst-fast.py
#!/usr/bin/env python3

def pig(first, last):
    return last[0] + first[1:]

wrong = 0
while True:
    full_name = input("What is your full name? ")
    split_name = full_name.split(' ')
    name_count = len(split_name)
    if full_name == ':q' or full_name == ":quit":
        break
    elif name_count != 2:
        wrong += 1
        print(f"{full_name} has {name_count} words instead of only 2: attempt {wrong}")
    else:
        first_name = split_name[0]
        last_name = split_name[1]
        lirst = pig(first_name, last_name)
        fast = pig(last_name, first_name)
        print(f"{lirst} {fast}")
        wrong = 0

File: Scala/lirst-fast/LirstFast.scala
import scala.annotation.tailrec
import scala.io.StdIn.readLine

object Main extends App {

  def pig(first: String, last: String): String = 
    (last.head +: first.tail).mkString

  def prompt: List[String] = 
    readLine("What is your full name? ").split(" ").toList

  final case class State(quit: Boolean, wrong: Int)

  def handle(state: State): State = 
    prompt match {
      case first :: last :: Nil =>
        val lirst = pig(first, last)
        val fast = pig(last, first)
        println(s"$lirst $fast")
        State(false, 0)
      case ":q" :: Nil | ":quit" :: Nil =>
        State(true, state.wrong)
      case badName@_ => 
        val name = badName.mkString(" ")
        val wordCount = badName.size
        val wrongInc = state.wrong + 1
        println(s"$name has $wordCount words instead of only 2: attempt ${wrongInc}")
        State(false, wrongInc)
    }

  @tailrec
  def run(prev: State, handle: State => State): State = {
    val next = handle(prev)
    if (!next.quit) run(next, handle)
    else next
  }

  run(State(false, 0), handle)
}
